distrobuilder/main.go

```diff
@@ -478,18 +478,137 @@ func getDefinition(fname string, options []string) (*shared.Definition, error) {
+# is_container succeeds if we're running inside a container
+is_container() {
+	grep -qa container=lxc /proc/1/environ
+}
+
+# Exit immediately if this is neither a VM nor a container
+if ! systemd-detect-virt && ! is_container; then
+	exit
+fi
+
+# Determine systemd version
+for path in /usr/lib/systemd/systemd /lib/systemd/systemd; do
+	[ -e "${path}" ] || continue
+	systemd_version="$("${path}" --version | head -n1 | cut -d' ' -f2)"
+	break
+done
+
+## Fixes
+
+# Fix capabilities
+(
+	if [ "${systemd_version}" -ge 244 ]; then
+		# Create top level dropin
+		dropin_dir="/run/systemd/system/service.d"
+		mkdir -p "${dropin_dir}"
+		echo "[Service]" > "${dropin_dir}/lxc.conf"
+		[ "${systemd_version}" -ge 247 ] && echo "ProtectProc=default" >> "${dropin_dir}/lxc.conf"
+		echo "ProtectControlGroups=no" >> "${dropin_dir}/lxc.conf"
+		echo "ProtectKernelTunables=no" >> "${dropin_dir}/lxc.conf"
+	else
+		# Create per-unit dropin for all service units
+		find /etc/systemd /run/systemd /usr/lib/systemd -name "*.service" -type f | sed -r 's#/usr/lib/systemd/##;s#/etc/systemd/##g;s#/run/systemd/##g' | while read service_file; do
+			dropin_dir="/run/systemd/${service_file}.d"
+			mkdir -p "${dropin_dir}"
+			echo "[Service]" > "${dropin_dir}/lxc.conf"
+			[ "${systemd_version}" -ge 247 ] && echo "ProtectProc=default" >> "${dropin_dir}/lxc.conf"
+			[ "${systemd_version}" -ge 232 ] && echo "ProtectControlGroups=no" >> "${dropin_dir}/lxc.conf"
+			[ "${systemd_version}" -ge 232 ] && echo "ProtectKernelTunables=no" >> "${dropin_dir}/lxc.conf"
+		done
+	fi
+)
+
+# Network fix for fedora/34/cloud
+(
+	. /etc/os-release
+
+	if ! which cloud-init || ! which NetworkManger || ! [ -e /sys/class/net/eth0 ] || ! is_container || [ "${ID}" != "fedora" ] || [ "${VERSION_ID}" != "34" ]; then
+		exit
+	fi
+
+	cat <<-EOF > /run/systemd/system/network-connection-activate.service
+[Unit]
+Description=Activate connection
+After=NetworkManager-wait-online.service
+
+[Service]
+ExecStart=-/usr/bin/nmcli c up "System eth0"
+Type=oneshot
+RemainAfterExit=true
+
+[Install]
+WantedBy=default.target
+EOF
+
+	mkdir -p /run/systemd/system/default.target.wants
+	ln -s /run/systemd/system/network-connection-activate.service /run/systemd/system/default.target.wants/network-connection-activate.service
+)
+
+# NetworkManager fix for various distributions
+(
+	if ! which NetworkManager || ! [ -f /bin/ip ] || ! [ -e /sys/class/net/eth0 ]; then
+		exit
+	fi
+
+	cat <<-EOF > /run/systemd/system/network-device-down.service
+[Unit]
+Description=Turn off network device
+Before=NetworkManager.service
+
+[Service]
+ExecStart=-/bin/ip link set eth0 down
+Type=oneshot
+RemainAfterExit=true
+
+[Install]
+WantedBy=default.target
+EOF
+
+	mkdir -p /run/systemd/system/default.target.wants
+	ln -s /run/systemd/system/network-device-down.service /run/systemd/system/default.target.wants/network-device-down.service
+)
+
+# Fix for systemd-networkd
+(
+	. /etc/os-release
+
+	[ "${ID}" = "altlinux" ] || exit
```

Comment:
I've got a feeling that this fix is directed towards new-ish systemd versions, but others were forgotten? At the very least Archlinux also needs (a superset of) this!

----

Actually, since systemd 244.1, this seems to be what's needed:
```ini
[Service]
BindReadOnlyPaths=/sys /proc
ProtectSystem=no
```
Not sure if ProtectSystem=no is acceptable to ship for a network-ish service, maybe that's a separate discussion to have.

Reply:
> At the very least Archlinux also needs (a superset of) this!
> Actually, since systemd 244.1, this seems to be what's needed to run unprived without(!) security.nesting=true: 

This is incorrect, at least for Arch Linux. `systemd-networkd` works fine there without `security.nesting=true`.

Reply:
~~That is absolutely not what i'm experiencing with `images:archlinux/cloud` on my debian buster host. Should I open a separate ticket?~~
Actually I've spun up a fresh `images:debian/buster/cloud` VM and installed lxd. In contrast to lxd on the host (:upside_down_face: ), `images:archlinux/cloud` does succeed in starting networkd in the VM's lxd, but no ipv4 until you add /sys to readonly like in this generator.
Sorry for the noise, guess I should've started there.